bagitobjecttransfer/recordtransfer/models.py

```diff
@@ -39,138 +42,364 @@ def get_full_name(self):
+        REMOVING_IN_PROGRESS = "RP", _("File Removal in Progress")
+        DELETED = "DL", _("All Files Removed")
+
     token = models.CharField(max_length=32)
     started_at = models.DateTimeField()
+    status = models.CharField(
+        max_length=2, choices=SessionStatus.choices, default=SessionStatus.CREATED
+    )
 
     @classmethod
-    def new_session(cls):
-        """Start a new upload session"""
+    def new_session(cls) -> "UploadSession":
+        """Start a new upload session."""
         return cls(token=get_random_string(length=32), started_at=timezone.now())
 
     @property
-    def upload_size(self):
-        """Get total size (in bytes) of all uploaded files in this session"""
-        return sum(f.file_upload.size for f in self.get_existing_file_set())
+    def upload_size(self) -> int:
+        """Get total size (in bytes) of all uploaded files in this session. This includes the size
+        of both temporary and permanent files.
+        """
+        if self.status in (
+            self.SessionStatus.EXPIRED,
+            self.SessionStatus.DELETED,
+        ):
+            raise ValueError(
+                f"""Cannot get upload size from session {self.token} because the session has
+                expired or has been deleted"""
+            )
+        elif self.status in (
+            self.SessionStatus.COPYING_IN_PROGRESS,
+            self.SessionStatus.REMOVING_IN_PROGRESS,
+        ):
+            raise ValueError(
+                f"""Cannot get upload size from session {self.token} while copying or removing
+                files is in progress"""
+            )
+        return sum(
+            f.file_upload.size
+            for f in chain(self.tempuploadedfile_set.all(), self.permuploadedfile_set.all())
+            if f.exists
+        )
+
+    @property
+    def file_count(self) -> int:
+        """Get the total count of temporary + permanent uploaded files."""
+        if self.status in (
+            self.SessionStatus.EXPIRED,
+            self.SessionStatus.DELETED,
+        ):
+            raise ValueError(
+                f"""Cannot get file count from session {self.token} because the session has
+                expired or has been deleted"""
+            )
+        elif self.status in (
+            self.SessionStatus.COPYING_IN_PROGRESS,
+            self.SessionStatus.REMOVING_IN_PROGRESS,
+        ):
+            raise ValueError(
+                f"""Cannot get file count from session {self.token} while copying or removing files
+                is in progress"""
+            )
+        return sum(
+            f.exists
+            for f in chain(self.permuploadedfile_set.all(), self.tempuploadedfile_set.all())
+        )
 
-    def get_existing_file_set(self):
-        """Get all files from the uploadedfile_set where the file exists"""
-        return [f for f in self.uploadedfile_set.all() if f.exists]
+    def add_temp_file(self, file: UploadedFile) -> "TempUploadedFile":
```

Comment:
Since we're going to include `from __future__ import annotations` at the top of this file, you can refer to the class name without quotes.

```suggestion
    def add_temp_file(self, file: UploadedFile) -> TempUploadedFile:
```

Reply:
Addressed by https://github.com/NationalCentreTruthReconciliation/Secure-Record-Transfer/pull/318/commits/8ace49c7759f1c1f5fb089f9d652b71a54497865